### ISSUE TITLE
Architecture guess for RT-DETR is broken

### ISSUE THREAD

**Issue by SteveHawk** [ultralytics/ultralytics#6452]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

Previously in `ultralytics/cfg/__init__.py`, the framework will automatically detect which model architecture to use, by checking if `rtdetr` is present in model's path:

https://github.com/ultralytics/ultralytics/blob/ec977f6ca7131dce985e4577e3ee7fffc8d2633b/ultralytics/cfg/__init__.py#L408-L411

However, in commit 87233ea, the mechanism changes to only checking the stem (filename without extension) of model's path:

https://github.com/ultralytics/ultralytics/blob/87233ea17c3286bfdd13b0616d878971ca43cef0/ultralytics/cfg/__init__.py#L408-L412

Unfortunately, the names of saved checkpoints are hard-coded to be `best.pt` and `last.pt`:

https://github.com/ultralytics/ultralytics/blob/6baa3bdde6b38285af1af2677e0fd8d0443008dd/ultralytics/engine/trainer.py#L97-L97

So RT-DETR can no longer be correctly loaded when using saved checkpoints, unless manually changing the name or code.

Please help fix this problem. In addition, this auto arch detection has always been confusing to use IMHO (also undocumented I believe), I suggest providing a new parameter to specify/override the intended model architecture to use, so we don't need to struggle with this obscure mechanism any more.

### Environment

```
Ultralytics YOLOv8.0.196 🚀 Python-3.11.4 torch-2.0.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
Setup complete ✅ (24 CPUs, 196.6 GB RAM, 75.0/98.9 GB disk)                                    
                                                                                                
OS                  Linux-5.4.0-139-generic-x86_64-with-glibc2.31                               
Environment         Linux                                                                       
Python              3.11.4                                                                      
Install             git                                                                         
RAM                 196.59 GB                                                                   
CPU                 Intel Xeon Gold 6226R 2.90GHz                                               
CUDA                11.8                                                                        
                                                                                                
matplotlib          ✅ 3.7.2>=3.3.0                                                             
numpy               ✅ 1.24.3>=1.22.2                                                           
opencv-python       ✅ 4.8.0.74>=4.6.0                                                          
pillow              ✅ 9.4.0>=7.1.2                                                             
pyyaml              ✅ 6.0>=5.3.1                                                               
requests            ✅ 2.31.0>=2.23.0                                                           
scipy               ✅ 1.11.1>=1.4.1                                                            
torch               ✅ 2.0.1>=1.8.0                                                             
torchvision         ✅ 0.15.2>=0.9.0                                                            
tqdm                ✅ 4.65.0>=4.64.0                                                           
pandas              ✅ 2.0.3>=1.1.4                                                             
seaborn             ✅ 0.12.2>=0.11.0                                                           
psutil              ✅ 5.9.5                                                                    
py-cpuinfo          ✅ 9.0.0                                                                    
thop                ✅ 0.1.1-2209072238>=0.1.1                                                  
```

### Minimal Reproducible Example

```
yolo detect predict \
    model=./rtdetr-runs/train/weights/best.pt \
    source=./dataset/test/test-images/ \
    project=rtdetr-runs \
    line_width=1 \
    device=0 \
    half=true
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@SteveHawk hello,

Thank you for bringing this to our attention and for the detailed explanation of the issue with RT-DETR architecture detection. Your insight into the change from a path-inclusive check to a filename-stem check and how it impacts the model loading is invaluable.

The intention behind changing the model checkpoint naming convention to a standard `best.pt` and `last.pt` was to simplify the checkpointing process. However, we understand that this, coupled with the change in architecture detection, has led to complications when working with RT-DETR.

The suggestion you've made about providing a clear parameter to specify or override the intended model architecture makes perfect sense. It would indeed offer a straightforward solution to users who want to load models without relying on naming conventions or specific detection mechanisms within the codebase.

We will work on addressing this issue and consider implementing a more explicit way to declare the architecture. We appreciate user feedback as it helps us improve the usability and functionality of the Ultralytics YOLO models.

In the meantime, as a temporary workaround, users may have to rename their RT-DETR model checkpoint files or adjust the code accordingly to ensure correct architecture detection.

We value your contribution to the YOLO community, and updates regarding this issue will be communicated through subsequent commits and releases in the YOLOv8 repository. If you would like to track the progress or contribute further to the resolution, please feel free to engage with the community on our issues page.

Warm

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR (#20578). This update introduces several improvements to model handling in the Ultralytics codebase:

- The Model class now supports already-initialized Model instances, not just file paths or names.
- The YOLO class constructor is enhanced to handle Model objects directly, streamlining workflows.
- Automatic detection and conversion of RTDETR models within the YOLO class makes integration smoother and requires fewer manual steps.
- General improvements to flexibility and user-friendliness for working with custom or pre-initialized models.

Special thanks to the PR author and contributors for these enhancements!

To test whether these changes resolve your issue, you can upgrade to the latest version (8.3.141) with:
```
pip install -U ultralytics>=8.3.141
```
Alternatively, you can install the main branch directly:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Feel free to wait for the next official PyPI release if preferred. Please let us know if this update resolves your issue or if you encounter any further problems. Thank you for reporting this and helping improve the project! 🙏
